packages/monaco/src/browser/monaco-editor-model.ts

```diff
@@ -158,49 +176,86 @@ export class MonacoEditorModel implements ITextEditorModel, TextEditorDocument {
         this.doAutoSave();
     }
 
     protected doAutoSave(): void {
         if (this.autoSave === 'on') {
+            const token = this.cancelSave();
             this.toDisposeOnAutoSave.dispose();
             const handle = window.setTimeout(() => {
-                this.doSave(TextDocumentSaveReason.AfterDelay);
+                this.scheduleSave(TextDocumentSaveReason.AfterDelay, token);
             }, this.autoSaveDelay);
             this.toDisposeOnAutoSave.push(Disposable.create(() =>
                 window.clearTimeout(handle))
             );
         }
     }
 
-    protected async doSave(reason: TextDocumentSaveReason): Promise<void> {
-        if (this.resource.saveContents) {
-            await this.fireWillSaveModel(reason);
-            if (this.dirty) {
-                const content = this.model.getValue();
-                await this.resource.saveContents(content);
-                this.setDirty(false);
-            }
-            this.onDidSaveModelEmitter.fire(this.model);
+    protected saveCancellationTokenSource = new CancellationTokenSource();
+    protected cancelSave(): CancellationToken {
+        this.saveCancellationTokenSource.cancel();
+        this.saveCancellationTokenSource = new CancellationTokenSource();
+        return this.saveCancellationTokenSource.token;
+    }
+
+    protected scheduleSave(reason: TextDocumentSaveReason, token: CancellationToken = this.cancelSave()): Promise<void> {
+        return this.run(() => this.doSave(reason, token));
+    }
+
+    protected async doSave(reason: TextDocumentSaveReason, token: CancellationToken): Promise<void> {
+        if (token.isCancellationRequested || !this.resource.saveContents) {
+            return;
+        }
+
+        await this.fireWillSaveModel(reason, token);
+        if (token.isCancellationRequested && !this.dirty) {
+            return;
         }
+
+        const content = this.model.getValue();
+        await this.resource.saveContents(content);
+        if (token.isCancellationRequested) {
+            return;
+        }
+
+        this.setDirty(false);
+        this.onDidSaveModelEmitter.fire(this.model);
     }
 
-    protected fireWillSaveModel(reason: TextDocumentSaveReason): Promise<void> {
+    protected fireWillSaveModel(reason: TextDocumentSaveReason, token: CancellationToken): Promise<void> {
         const model = this.model;
         return new Promise<void>(resolve => {
             this.onWillSaveModelEmitter.fire({
                 model, reason,
                 waitUntil: thenable =>
                     thenable.then(operations => {
-                        model.applyEdits(operations);
+                        if (token.isCancellationRequested) {
+                            resolve();
+                        }
+                        const invertedOperations = this.applyEdits(operations);
+                        token.onCancellationRequested(() => this.applyEdits(invertedOperations));
```

Comment:
Could you have the following sequence:

  -  await this.fireWillSaveModel(reason, token);
  -  await this.resource.saveContents(content);
  -  cancelSave()
  -  token.onCancellationRequested(() => this.applyEdits(invertedOperations)); executes
  - if (token.isCancellationRequested) return true

And this way saveContents will have saved the contents with the result from `this.applyEdits(operations);`

but what will be shown to the user is `this.applyEdits(invertedOperations));`

Reply:
It's a bit hard to reason about it since there are no clients. Thinking about it probably we should never rollback changes did in `fireWillSaveModel `. Otherwise, a user will see appearing and disappearing of changes, as well as can lose some changes, like if he types between `fireWillSaveModel` and `saveContents`.

I can comment out `fireWillSaveModel` and we handle it separately?

Reply:
Yes seems like a good idea to me.

Sorry for the delay I missed that message.

Reply:
I've removed operations reverting, so if a user types something after applying operations then changes won't be lost.